Update messaging "Receive", "Deliver", and "Create" operations according to OTEP 220 (#284)
open-telemetry/semantic-conventions: docs/messaging/messaging-spans.md

* `AuthenticationRequest-Conversations process`
* `(anonymous) publish` (`(anonymous)` being a stable identifier for an unnamed destination)

### Operation names

The following operations related to messages are defined for these semantic conventions:

| Operation name | Description |
| -------------- | ----------- |
| `publish`      | One or more messages are provided for publishing to an intermediary. |
| `create`       | A message is created. |
| `receive`      | One or more messages are requested by a consumer. |
| `deliver`      | One or more messages are passed to a consumer. |

### Span kind

[Span kinds](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind)
SHOULD be set according to the following table, based on the operation a span describes.

| Operation name | Span kind|
|----------------|-------------|

> | `publish`      | `PRODUCER`, if no `create` spans are present. |

Reviewer: What should the Span kind be if create spans are present?

Author: Proposal from today's SIG meeting:
> `PRODUCER`, if the context of the publish span is used as creation context.

And add this elaboration from the OTEP:
> For cases not covered by the table above, the span kind should be set according to the [generic specification about span kinds](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind), e. g. it should be set to CLIENT if the publish spans models a synchronous call to the intermediary.
>
> Setting span kinds according to this table ensures that span links between consumers and producers always exist between a PRODUCER span on the producer side and a CONSUMER span on the consumer side. This allows analysis tools to interpret linked traces without the need for additional semantic hints.

> | `create`       | `PRODUCER`, |

Reviewer 2: nit 
```suggestion
| `create`       | `PRODUCER` |
```

| `receive`      | `CONSUMER` |
| `deliver`      | `CONSUMER` |

### Trace structure

#### Producer spans

"Publish" spans SHOULD be created for operations of providing messages for
sending or publishing to an intermediary. A single "Publish" span can account
for a single message, or for multiple messages (in the case of providing

> messages in batches). "Create" spans MAY be created. A single "Create" span

Reviewer 2: "Create" spans MAY be created." seems redundant

Author: Creating a "Publish" span is a strong requirement, it needs to present in all producer scenarios. We formulated it as a SHOULD.

Creating "Create" spans is a weaker requirement. It shouldn't be created for single-message scenarios and for scenarios where a creation context is provided. That's why we settled on a MAY, if I remember correctly.

Here's the [relevant part from the OTEP](https://github.com/open-telemetry/oteps/blob/main/text/trace/0220-messaging-semantic-conventions-span-structure.md#producer):

> There are four different scenarios for injecting a creation context into a message:
>
> 1. A user provides custom creation contexts for the messages that are published, which don't refer to any spans described in this document. This provides flexibility to users to model custom scenarios. In this case, no other additional spans besides the "Publish" span should be created. The "Publish" span should link to the provided creation contexts.
> 2. If no custom creation context is provided for a message, it is recommended to create a "Create" span for every single message and inject its context into the message. "Create" spans can be created during the "Publish" operation as children of the "Publish" span.
> 3. As a variation of the scenario above, "Create" spans can be created independently of the "Publish" operation, e. g. in cases where messages are created before they are passed to a "Publish" operation. In this case, the "Publish" span should link to the "Create" spans.
> 4. For single-message scenarios or when large number of spans are a problem, the context of the "Publish" span can be injected into the message, thus acting as the creation context. In this case, no other spans besides the "Publish" span should be created.

Author: Discussed in today's SIG meeting: try to explain operations in more detail in the operation names table (in the description).

SHOULD account only for a single message. "Create" spans SHOULD either be
children or links of the related "Publish" span.

If a "Create" span exists for a message, its context SHOULD be injected into
the message. If no "Create" span exists, the context of the related "Publish"
span SHOULD be injected into the message.

#### Consumer spans

"Deliver" spans SHOULD be created for operations of passing messages to the
application when those operations are not initiated by the application code
(push-based scenarios).

"Receive" spans SHOULD be created for operations of passing messages to the
application when those operations are initiated by the application code
(pull-based scenarios).

"Deliver" or "Receive" spans MUST NOT be created for messages that are
pre-fetched or cached by messaging libraries or SDKs until they are forwarded
to the caller.

A single "Deliver" or "Receive" span can account for a single message, for
multiple messages (in case messages are passed for processing as batches), or
for no message at all (if it is signalled that no messages were received). For
each message it accounts for, the "Deliver" or "Receive" span SHOULD link to

> the message's creation context. In addition, if it is possible the creation

Author: I'd prefer keeping implementation-specific details out of this document. But if you come across a formulation that fits, please propose a change.

> the message's creation context. In addition, if it is possible the creation

Reviewer 3: Read it again after our discussions and I think it's good as is :)

> the message's creation context. In addition, if it is possible the creation

Reviewer 4: Although rather informal/verbose in the OTEP itself, I'd like to get some justification/guideline on this one. From the OTEP:

> For single-message scenarios, and if the "Deliver" or "Receive" spans would be root spans of a new trace, the creation context may also be used as a parent on those operations in addition to being added as a link. Keeping single-messages operations in the same trace can greatly improve the user experience.

Otherwise this may be used as de facto backdoor to not use `Links` (better be safe than sorry).

Author: @carlosalberto 

> Otherwise this may be used as de facto backdoor to not use Links (better be safe than sorry).

We always require links to be present. Parent-child relationship can be created _in addition_, where it's possible.

I'd prefer to not elaborate too much on the parent-child relationship option (as I fear that might lead people astray), but I'm happy to add clarifications if you think it's necessary.

| `publish`      | One or more messages are provided for publishing to an intermediary. If a single message is published, the context of the "Publish" span can be used as the creation context and no "Create" span needs to be created. |
| `create`       | A message is created. "Create" spans always refer to a single message and are used to provide a unique creation context for messages in batch publishing scenarios. |
| `receive`      | One or more messages are requested by a consumer. This operation refers to pull-based scenarios, where consumers explicitly call methods of messaging SDKs to receive messages. |
| `deliver`      | One or more messages are passed to a consumer. This operation refers to push-based scenarios, where consumer register callbacks which get called by messaging SDKs. |

### Span kind

[Span kinds](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind)
SHOULD be set according to the following table, based on the operation a span describes.

| Operation name | Span kind|
|----------------|-------------|
| `publish`      | `PRODUCER`, if the context of the "Publish" span is used as creation context. |

> | `create`       | `PRODUCER`, |

Reviewer: Minor point, but I think the trailing ',' should be removed here.

| `receive`      | `CONSUMER` |
| `deliver`      | `CONSUMER` |

For cases not covered by the table above, the span kind should be set according
to the [generic specification about span kinds](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind),
e. g. it should be set to CLIENT for the "Publish" span if its context is not
used as creation context and if the "Publish" span models a synchronous call to
the intermediary.

> Setting span kinds according to this table ensures that span links between

Reviewer 3: Does "span links" here does refer to actual links or more of a "connection" between them? If it's the latter, I think we may need to remove the word "link" here to avoid confusion.

Author: I'm inclined to leave the reference to "links" in place, as it's indeed a link that we guarantee between producer and consumer spans. There might or might not be a parent-child relationship in addition (we're very opaque about that now as we removed any reference to it), but there will always be a link.

consumers and producers always exist between a PRODUCER span on the producer
side and a CONSUMER span on the consumer side. This allows analysis tools to
interpret linked traces without the need for additional semantic hints.

### Trace structure

#### Producer spans

"Publish" spans SHOULD be created for operations of providing messages for
sending or publishing to an intermediary. A single "Publish" span can account
for a single message, or for multiple messages (in the case of providing
messages in batches). "Create" spans MAY be created. A single "Create" span
SHOULD account only for a single message. "Create" spans SHOULD either be
children or links of the related "Publish" span.

If a "Create" span exists for a message, its context SHOULD be injected into
the message. If no "Create" span exists and no custom creation context is
injected into the message, the context of the related "Publish" span SHOULD be

> injected into the message.

Reviewer 2: I think we completely lost this point from the OTEP
> A user provides custom creation contexts for the messages that are published, which don't refer to any spans described in this document. This provides flexibility to users to model custom scenarios. In this case, no other additional spans besides the "Publish" span should be created. The "Publish" span should link to the provided creation contexts.

It's also not reflected in the [Context Propagation](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/messaging/messaging-spans.md#context-propagation) section.

Author: @lmolkova What do you think about this wording:

> If a user provides a custom creation context for a message, this context SHOULD be injected into the message.  Otherwise, if a "Create" span exists for a message, its context SHOULD be injected into the message. If no "Create" span exists and no custom creation context is injected into the message, the context of the related "Publish" span SHOULD be injected into the message.

Reviewer 2: The part that's missing that Publish operation should be linked to the message context.

I'd add something g like 

"If a user provides a custom creation context int a message, this context SHOULD NOT be modified: `Create` span  SHOULD NOT be created. The custom creation context should be linked to the `Publish` span."

I.e.:
- when publishing/creating a message:
  - extract context. 
     - if context already exists, add it as a link to Publish span
     - otherwise, start publish/create span, inject context

Author: @lmolkova Got it, I ended up changing the wording to this:

> If a user provides a custom creation context in a message, this context SHOULD NOT be modified, a "Create" span SHOULD NOT be created, and the "Publish" span SHOULD link to the custom creation context.  Otherwise, if a "Create" span exists for a message, its context SHOULD be injected into the message. If no "Create" span exists and no custom creation context is injected into the message, the context of the related "Publish" span SHOULD be injected into the message.

Let me know if you think that covers the whole custom creation context scenario.

#### Consumer spans

"Deliver" spans SHOULD be created for operations of passing messages to the
application when those operations are not initiated by the application code
(push-based scenarios).

> "Receive" spans SHOULD be created for operations of passing messages to the

Reviewer 5: For a system that polls but finds no messages, do you get a Receive Span / Independent Trace?

Sorry if this is a silly question, trying to understand how you would be observing the cost of polling, e.g. how many times you try to receive but find no messages available.  Perhaps that's handled on the metrics side?

Author: Good question, the answer is two times "yes".

1. Yes, it's planned to have metrics that cover this: https://github.com/open-telemetry/semantic-conventions/pull/163/files#diff-1c3afdf1b3b840fc9e4c17a2110573d1656b7e27d75b13f4ce97e2daff22517bR13-R14

2. Yes, if you initiate a pull-operation and you don't receive any message, you should still get a span. See this text two paragraphs lower:

> A single "Deliver" or "Receive" span can account for a single message, for a batch of messages, or for no message at all (if it is signalled that no messages were received).

SHOULD account only for a single message. "Create" spans SHOULD either be
children or links of the related "Publish" span.

If a user provides a custom creation context in a message, this context SHOULD
NOT be modified, a "Create" span SHOULD NOT be created, and the "Publish" span
SHOULD link to the custom creation context.  Otherwise, if a "Create" span
exists for a message, its context SHOULD be injected into the message. If no
"Create" span exists and no custom creation context is injected into the
message, the context of the related "Publish" span SHOULD be injected into the
message.

#### Consumer spans

"Deliver" spans SHOULD be created for operations of passing messages to the
application when those operations are not initiated by the application code
(push-based scenarios).

"Receive" spans SHOULD be created for operations of passing messages to the
application when those operations are initiated by the application code
(pull-based scenarios).

"Deliver" or "Receive" spans MUST NOT be created for messages that are
pre-fetched or cached by messaging libraries or SDKs until they are forwarded
to the caller.

> A single "Deliver" or "Receive" span can account for a single message, for a

Reviewer 2: I think we're missing something that talks about deliver span duration from the OTEP: "A "Deliver" span covers the call of such a callback or handler"

Author: I added this to the first paragraph:

> "Deliver" spans SHOULD be created for operations of passing messages to the application when those operations are not initiated by the application code (push-based scenarios). _A "Deliver" span covers the duration of such an operation, which is usually a callback or handler._

| `receive`      | A message is received from a destination by a message consumer/server. |
| `process`      | A message that was previously received from a destination is processed by a message consumer/server. |

### Span kind

[Span kinds](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind)
SHOULD be set according to the following table, based on the operation a span describes.

| Operation name | Span kind|
|----------------|-------------|
| `publish`      | `PRODUCER` |

> | `receive`      | `CONSUMER` |

Reviewer 2: we should probably either remove process now or add it to the table here since this section is used to vaguely cover it.

Author: In the messaging WG meeting two weeks ago we decided to incorporate OTEP-proposed changes for `create` and `deliver` span into this PR, because otherwise we'll end up with a barely usable intermediate state (e. g. ambiguity about the process operation, or also the fact that we'll only support pull-based scenarios until `deliver` is added too).

the broker does not have such notion, the original destination name SHOULD uniquely identify the broker.
<!-- endsemconv -->

"Receive" spans SHOULD be created for operations of passing messages to the application when those operations are initiated by the application code.

> "Receive" spans MUST NOT be created for messages which are not forwarded to the caller, but are pre-fetched or cached by messaging libraries or SDKs.

Reviewer 2: ```suggestion
"Receive" spans MUST NOT be created for messages until they are forwarded to the caller, but are pre-fetched or cached by messaging libraries or SDKs.
```

otherwise it's possible to interpret it as that spans should never be created

### Topic with multiple consumers

> Given is a process P, that publishes a message to a topic T on messaging system MS, and two processes CA and CB, which both receive the message.

Reviewer 3: The text now does not align anymore with the new diagram (there's no CA/CB in the mermaid). Would be good if they match, otherwise might be confusing to folks.

Author: Fixed.

```mermaid
flowchart LR;
  subgraph PRODUCER
  direction TB
  P[Span Publish A]
  end
  subgraph CONSUMER1
  direction TB
  R1[Span Receive A 1]
  end
  subgraph CONSUMER2
  direction TB
  R2[Span Receive A 2]
  end
  P-. link .-R1;
  P-. link .-R2;

  classDef normal fill:green
  class P,R1,R2 normal
  linkStyle 0,1 color:green,stroke:green
```

| Field or Attribute | Span Publish A | Span Receive A 1| Span Receive A 2 |
|-|-|-|-|
| Span name | `T publish` | `T receive` | `T receive` |
| Parent | | | |
| Links |  | `T publish` | `T publish` |
| SpanKind | `PRODUCER` | `CONSUMER` | `CONSUMER` |
| Status | `Ok` | `Ok` | `Ok` |
| `server.address` | `"ms"` | `"ms"` | `"ms"` |
| `server.port` | `1234` | `1234` | `1234` |
| `messaging.system` | `"rabbitmq"` | `"rabbitmq"` | `"rabbitmq"` |
| `messaging.destination.name` | `"T"` | `"T"` | `"T"` |

> | `messaging.operation` |  | `"receive"` | `"receive"` |

Reviewer 2: ```suggestion
| `messaging.operation` | `"publish"` | `"receive"` | `"receive"` |
```

?

| `messaging.message.id` | `"a"` | `"a"`| `"a"` |

### Batch delivering

Given is a process P, that publishes two messages to a queue Q on messaging system MS, and a process C, which gets both of them delivered in one batch (Span Recv1) and processes each message separately.

> Since a span can only have one parent, the `deliver` span will have no parent will be correlated with the producing spans using links.

Reviewer 3: ```suggestion
Since a span can only have one parent, the `deliver` span will have no parent and will be correlated with the producing spans using links.
```

### Topic with multiple consumers

> Given is a publisher that publishes a message to a queue "T" on RabbitMQ, and two consumers which both receive the message.

Reviewer: The text being replaced describes a scenario where a publisher is publishing to a topic T, but this is replaced with publishing to a queue T. I think publishing to a topic here makes much more sense than publishing to a queue. A queue is a point-to-point communication mechanism whereas a topic is intended to reach multiple consumers. I'll concede it's usually technically possible for multiple consumers to receive a message from a queue, but this is only in corner cases where, perhaps, a consumer receives the message and is then disconnected before it is able to acknowledge the message. I think it would be better to illustrate multiple consumers with a normal scenario rather than a corner case scenario.

There may be a more definitive source of information, but this reference shows how publishing to a RabbitMQ topic would be received by multiple consumers:
https://www.rabbitmq.com/tutorials/tutorial-five-python.html

Author: You're right, things got mixed up here. Also, the RabbitMQ example had the title "Topic with multiple consumers", which is confusing.

I re-wrote the examples as follows:
* "Topic with multiple consumers": Kafka, topic, push-based consumption (deliver).
* "Batch receiving": RabbitMQ, queue, pull-based consumption (receive).

Reviewer: I'm not sure RabbitMQ is the best messaging system to use to illustrate batch message processing. I'm not a RabbitMQ expert, but in a quick search I wasn't able to see how batch messaging could be performed with "out of the box" RabbitMQ. I did find batching examples where RabbitMQ is used under the hood , but in all cases I found, this is where the batching was performed by some software on top of the underlying RabbitMQ messaging system (ex. Spring Cloud Streams, Spring Boot, and various home grown layers or application workarounds).

In the [RabbitMQ docs](https://www.rabbitmq.com/documentation.html) and one of the client guides ([Java Client](https://www.rabbitmq.com/api-guide.html)), I wasn't able to find any explanation how to send or receive with batches.

I think it may be preferable to continue using RabbitMQ as the example publishing on topic T to multiple consumers, and continue to use Kafka as the batch consumer example from topic T.

Author: Ok, I ended up with those two scenarios now:

* "Topic with multiple consumers": RabbitMQ, topic exchange, push-based consumption (deliver).
* "Batch receiving": Kafka, topic, pull-based consumption (receive).

Reviewer: This looks good to me, thanks @pyohannes.
